docs/bazel/development.md

```diff
@@ -0,0 +1,53 @@
+# development
+
+* To get more verbose output use the flags documented in `.bazelrc`.
+* For convenience, use a `user.bazelrc` file (but do not place it
+  under version control).  Enable the following (which you can copy
+  from the `.bazelrc` file):
+  * build --color=yes
+  * build --subcommands    # prints build cmd with space-separate args
+  * build --subcommands=pretty_print  # same, but args are newline-separated
+  * build --verbose_failures
+  * build --sandbox_debug
+* For more details on `rc` files, see [.bazelrc, the Bazel configuration file](https://docs.bazel.build/versions/master/guide.html#bazelrc-the-bazel-configuration-file).  See also [Best Practices - .bazelrc](https://docs.bazel.build/versions/master/best-practices.html#bazelrc)
+* Bazel uses a bunch of hidden files and settings. To list them run `$ bazel info`.
+  * You can also pass one of the keys as parameter, e.g. `$ bazel info
+    output_base`.  You can use this in shell scripts, for example, to
+    make your own convenience tools.
+  * External repos populate `$(bazel info output_base)/external`.  But
+    note that it will only be populated when you buld a target that
+    depends on the external repo.  So if you want to e.g. see what the
+    `@digestif` repo looks like to Bazel, first build one of its
+    targets and then do something like `$ ls $(bazel info output_base)/external/digestif`.
+  * See [Output directory
+    layout](https://docs.bazel.build/versions/master/output_directories.html)
+    for more details.
+* Bazel creates softlinks in your root repo dir pointing to some of
+  its hidden directories:
+  * `bazel-bin`
+  * `bazel-out`
+  * `bazel-testlogs`
+  * `bazel-<root>` - for example, if your base WORKSPACE files is in dir foo, this will be `bazel-foo`
+
+  If you find having these directories in the top level inconvenient,
+  you can tell Bazel to put them somewhere else by passing `--symlink_prefix`.  A common practice is to put the following line in your `user.bazelrc` file, in order to put them in a hidden dot-dir:
+  * build --symlink_prefix=.bazel/
+* Bazel keeps a log of each command in `bazel info command_log`, so if
+  you miss the output you can find it there.  But note that each
+  command overwrites it - including the `info` command.  So if you run
+  a command and then use `bazel info command_log` to find the log, it
+  will contain the result of running `bazel info command_log`. Ha ha ha!
+  So you need to save the name of the log somewhere.
+* If you are working on an external repository, you can tell Bazel to
```

Comment:
Not sure what you mean.  Currently external repos are embedded as git submodules.  The bazel repo rules will work with this, they don't know anything about git, they just need a path to a repo.  Using repo rules means we have the option of not embedding such code.  There's no real reason to do so, technically.  Unless you need to edit the code, an external repo is just another library, we can leave it to bazel to download and compile in its own cache. If you need to edit the code, you need to download it, and you need to tell Bazel to use your local copy, which is easy with --override_repository.  That's how I've been testing stuff, it works very well, as long as you remember set/unset the param.

Reply:
> Using repo rules means we have the option of not embedding such code. There's no real reason to do so, technically.

I agree, but @bkase is a strong believer in monorepos, and I think you're going to have a hard time persuading him to punt the submodules completely out-of-tree.

As it stands, I think we should always use the submodule's path directly rather than an external repo reference. Having submodules potentially out of sync with bazel is confusing and an accident waiting to happen.

Reply:
Works either way for me, those are policy decisions for you folks.  I'm not sure there's a strong technical reason to prefer one over the other.

Reply:
I am a strong believer in monorepos with no submodules, once we have submodules, the world is already not happy and I think it does actually make sense to rip them out-of-tree. But I think we can only do this when we're done supporting backwards compat with dune, right?

Reply:
Yep.  Probably don't want to do any major surgery until we're completely confident in the Bazel builds.